Delete file on rollback in HiveRecordWriter

losipiuk opened the pull request:
Currently if table write operation was interrupted by user
the files which were already partly written to table directory were not
deleted.
In RecordFileWriter.rollback() the RecordWriter.close() was called with
abort parameter set to true. But implementations of RecordWriters ignore
that parameter.
That resulted with garbage files for CTAS/INSERT flows which did not use
temporary directory (when S3 FS was in use).

By explicitly deleting file in RecordFileWriter.rollback() we make this
proble less probable to surface.

pnowojski commented on presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java:
```diff
@@ -161,6 +163,8 @@ public void rollback()
         try {
             recordWriter.close(true);
+            // perform explicit delete of written file here as recordWriter.close() ignore the abort flag.
+            path.getFileSystem(conf).delete(path, false);
```
What will happen if the created file is not yet visible? I wasn't able to find an answer to this question here:

http://docs.aws.amazon.com/AmazonS3/latest/dev/Introduction.html#ConsistencyModel

but it states that: `A process writes a new object to Amazon S3 and immediately lists keys within its bucket. Until the change is fully propagated, the object might not appear in the list. ` However there is no explanation what happens with "delete-after-write" :/

losipiuk replied:
I don't know. When I tested that it seems to behave correctly - the files are deleted - even though delete is issed immediatly after closing file outputstream. But there may be some corner cases.

electrum replied:
Grammar nit: ignore -> ignores

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java:
```diff
@@ -77,6 +78,7 @@ public RecordFileWriter(
             TypeManager typeManager)
     {
         this.path = requireNonNull(path, "path is null");
+        this.conf = requireNonNull(conf, "conf can not be null");
```
nit: `conf is null`

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java:
```diff
@@ -161,6 +163,8 @@ public void rollback()
+            // perform explicit delete of written file here as recordWriter.close() ignore the abort flag.
+            path.getFileSystem(conf).delete(path, false);
         }
         catch (IOException e) {
```
What if `close()` succeeds and `delete()` throws `IOException`, do we still want to throw `HIVE_WRITER_CLOSE_ERROR`?

electrum replied:
I think throwing an exception here is wrong, since `rollback()` is called in a loop in `HivePageSink.doAbort()` to rollback all the writers. This means we can leak file descriptors or other resources.

losipiuk replied:
But `recordWriter.close()` can throw exception anywat, so if there is issue with lost resources, we have the problem already  in the current codebase.

Or do you suggest to remove the throwing of exception from `catch` clause in `rollback` at all?

electrum commented:
Please make the same change in `RcFileFileWriter`

losipiuk commented:
> Please make the same change in RcFileFileWriter

I would skip that unless @dain confirms that he has not code for that alrady in flight.
It is marked as todo and it seems Dain is currently working on that.

losipiuk commented:
added grammar fixups

electrum commented:
I think we should remove the throw and replace it with log warning. I do believe the existing code can leak.

losipiuk commented:
> I think we should remove the throw and replace it with log warning. I do believe the existing code can leak.

I added commit which replaces `throw` witl `log.warn`.

Yet it seems to me that it would be better to change the loop which calls rollback in `HivePageSink.doAbort()` to accumulate potenial exceptions thrown by `rollback` calls.
And in case any of the `rollback` calls threw an exception, rethrow a composite one.

This way we do not implement internal of `rollback` method to match way how it is called much deeper on the stack.
And we handle both `RecordFileWriter` and `RcFileWriter` with one logic.

What do you think?

electrum commented:
The new RC writer is all checked in and ready to use. The todo was a "this should be fixed" rather than "work in progress".

electrum commented:
Doing it in HivePageSink sounds good. I like that approach.

nezihyigitbasi commented:
I guess the same problem also applies to the `commit()` method, throwing in that method may result in a leak.

losipiuk commented:
Upadated patch.
Added file deleting to `RcFileWriter` and code which prevents breaking the loop of `rollback` calls if one of theses throws an exception.

I also changed original commit to perform delete in `finally` section even if `RecordWriter.close` throws an exception.

Per @nezihyigitbasi comment:

> I guess the same problem also applies to the commit() method, throwing in that method may result in a leak.

Isn't it mitigated by the fact that in case of exception on `commit()` rollback loop would be called afterwards? At least it is what I expect.

nezihyigitbasi commented:
@losipiuk You are probably right. I see rollback being called if commit doesn't go through in `TableWriterOperator::close()`.

losipiuk commented:
@nezihyigitbasi, @electrum please take a look when you have a chance :)

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java:
```diff
@@ -160,7 +162,13 @@ public void commit()
+                recordWriter.close(true);
+            }
+            finally {
+                // perform explicit delete of written file here as recordWriter.close() ignores the abort flag.
```
perform explicit deletion here as implementations of RecordWriter.close() often ignore the abort flag.

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/RcFileFileWriterFactory.java:
```diff
@@ -139,8 +140,14 @@ else if (ColumnarSerDe.class.getName().equals(storageFormat.getSerDe())) {
                 });
             }
 
+            Callable<Void> outputFileDeleteRoutine = () -> {
```
why don't you make this runnable and get rid of the return?

nezihyigitbasi replied:
you can also put the try catch here and remove it from `RcFileFileWriter.java` where you ignore it.

haozhun replied:
@losipiuk 
I had a discussion with @nezihyigitbasi. We agreed that having try...catch where you put it is better. Please ignore the last comment about moving the try catch.

haozhun replied:
`Callable` looks better.

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/RcFileFileWriter.java:
```diff
@@ -46,12 +47,14 @@
 
     public RcFileFileWriter(
             OutputStream outputStream,
+            Callable<Void> outputFileDeleteRoutine,
```
how about renaming this to `rollbackAction` or sth similar?

nezihyigitbasi commented on presto-hive/src/main/java/com/facebook/presto/hive/RcFileFileWriter.java:
```diff
@@ -136,10 +145,14 @@ public void commit()
+                outputFileDeleteRoutine.call();
+            }
         }
-        catch (IOException e) {
```
why did you change this?

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java:
```diff
@@ -212,12 +216,24 @@ public void abort()
+            exceptions.entrySet().forEach(
+                    entry -> log.warn("exception '%s' while rollback on %s", entry.getValue(), entry.getKey())
+            );
+            throw new PrestoException(HIVE_WRITER_CLOSE_ERROR, "Error rolling back write to Hive");
```
Put one of the exceptions as the cause for this. It would be nice to have the file name for that one as well.

@dain, @electrum, and I considered putting all of them in suppressed, and decided not to that. But we feel like that is not likely going to give more information. And when we add a list to suppressed, we have to be careful about not having too many of them, which is additional hassle.

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java:
```diff
@@ -212,12 +216,24 @@ public void abort()
         }
+        if (!exceptions.isEmpty()) {
+            exceptions.entrySet().forEach(
+                    entry -> log.warn("exception '%s' while rollback on %s", entry.getValue(), entry.getKey())
```
Why not print the logs as you encounter that? The list doesn't appear to be necessary.

losipiuk commented:
@haozhun back to you :)

losipiuk commented:
@haozhun  Any update on this one? It is small so let's get rid of it :)

haozhun commented:
Merged #7468

losipiuk commented:
Thanks

On Mar 25, 2017 01:49, "Haozhun Jin" <notifications@github.com> wrote:

> Closed #7468 <https://github.com/prestodb/presto/pull/7468>.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prestodb/presto/pull/7468#event-1015177912>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA2U_768pUi3Pg_pgPxoTT5zFGx07vUQks5rpGStgaJpZM4MOrof>
> .
>
